File: engine/src/board.cpp
#include "board.h"

#include <iostream>

int Board::drop(int col, Cell player_color) {
  if (col < 0 || col >= kCols) {
    return -1;
  }
  int row = getFirstEmptyRow(col);
  if (row == -1) {
    return -1;
  }
  cells_[row][col] = player_color;
  return row;
}

Cell Board::at(int row, int col) const { return cells_[row][col]; }

bool Board::isFull(int col) const { return cells_[0][col] != Cell::Empty; }

bool Board::isBoardFull() const {
  for (int c = 0; c < kCols; ++c) {
    if (!isFull(c)) {
      return false;
    }
  }
  return true;
}

void Board::print() const {
  for (int r = 0; r < kRows; ++r) {
    for (int c = 0; c < kCols; ++c) {
      char ch = '.';
      if (cells_[r][c] == Cell::Red) {
        ch = 'R';
      } else if (cells_[r][c] == Cell::Yellow) {
        ch = 'Y';
      }
      std::cout << ch << ' ';
    }
    std::cout << '\n';
  }
  std::cout << "0 1 2 3 4 5 6\n";
}

int Board::getFirstEmptyRow(int col) const {
  if (isFull(col)) {
    return -1;
  }
  for (int i = kRows - 1; i >= 0; --i) {
    if (cells_[i][col] == Cell::Empty) {
      return i;
    }
  }
  return -1;
}

File: server/internal/game/board.go
package game

import (
	"errors"
	"fmt"
)

var (
	ErrInvalidColumn = errors.New("invalid column")
	ErrColumnFull    = errors.New("column full")
)

type Board struct {
	cells [Rows][Cols]Cell
}

// Drop places a piece with gravity. Returns the row it landed in.
func (b *Board) Drop(col int, color Cell) (int, error) {
	if col < 0 || col >= Cols {
		return -1, ErrInvalidColumn
	}
	row := b.getFirstEmptyRow(col)
	if row == -1 {
		return -1, ErrColumnFull
	}
	b.cells[row][col] = color
	return row, nil
}

func (b *Board) At(row, col int) Cell {
	return b.cells[row][col]
}

// IsFull reports whether the top cell of the column is occupied.
func (b *Board) IsFull(col int) bool {
	return b.cells[0][col] != Empty
}

func (b *Board) IsBoardFull() bool {
	for c := 0; c < Cols; c++ {
		if !b.IsFull(c) {
			return false
		}
	}
	return true
}

func (b *Board) Cells() [Rows][Cols]Cell {
	return b.cells
}

func (b *Board) Print() {
	for r := 0; r < Rows; r++ {
		for c := 0; c < Cols; c++ {
			ch := "."
			switch b.cells[r][c] {
			case Red:
				ch = "R"
			case Yellow:
				ch = "Y"
			}
			fmt.Print(ch, " ")
		}
		fmt.Println()
	}
	fmt.Println("0 1 2 3 4 5 6")
}

// getFirstEmptyRow finds the lowest empty row in col (row 0 = top).
// Returns -1 if the column is full.
func (b *Board) getFirstEmptyRow(col int) int {
	if b.IsFull(col) {
		return -1
	}
	for i := Rows - 1; i >= 0; i-- {
		if b.cells[i][col] == Empty {
			return i
		}
	}
	return -1
}
